Regression when loading NuGet in MSBuild tasks

### Issue Description
https://github.com/dotnet/msbuild/pull/6126 pins NuGet loads from tasks to use the one in the SDK directly regardless of the version referenced by tasks.  This will unnaturally pin NuGet to the SDK version which will cause issues when tasks need to move ahead either for testing or for general functionality.  In the past we've been very careful about pinning NuGet.

Additionally, the closure of NuGet's assemblies is not loaded from ALC.  This leads to type unification problems, since the first load from MSBuild will cause NuGet to bind to these assemblies in MSBuild, then a task which runs later will bind to it's local copy (even if it is the exact same version).  https://github.com/dotnet/runtime/pull/48462#issuecomment-782854316
```
.packages/microsoft.dotnet.build.tasks.packaging/6.0.0-beta.21120.1/build/Packaging.targets(747,5): error MSB4018: (NETCORE_ENGINEERING_TELEMETRY=Build) The "GenerateRuntimeDependencies" task failed unexpectedly.
System.MissingMethodException: Method not found: 'Void NuGet.RuntimeModel.JsonObjectWriter..ctor(Newtonsoft.Json.JsonWriter)'.
   at Microsoft.DotNet.Build.Tasks.Packaging.NuGetUtility.WriteRuntimeGraph(String filePath, RuntimeGraph runtimeGraph)
   at Microsoft.DotNet.Build.Tasks.Packaging.GenerateRuntimeDependencies.Execute() in /_/src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeDependencies.cs:line 144
   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask)
```

cc @ViktorHofer 

### Steps to Reproduce
1. Run a task which uses both NuGet and Newtonsoft.JSON and calls NuGet APIs which use Newtonsoft.JSON types.  EG: https://github.com/dotnet/arcade/blob/ca7fab569267ed3bc73360882d652d119aae5653/src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeDependencies.cs


### Expected Behavior
Task runs to completion.

### Actual Behavior
Task fails with MissingMethodException since the calling task assembly and NuGet don't agree on the assembly that Newtonsoft.Json.JsonWriter lives in.

### Analysis
I think we need to rethink the solution for https://github.com/dotnet/msbuild/pull/6126.  I don't think the host should be pinning assemblies which don't need to unify.

It should be *possible* for tasks to defer to MSBuild's copy of assemblies if that works well for them.  In those cases the task should omit the entire assembly closure graph for those assemblies from their task layout, and they should ensure that they reference a version that is <= the one in MSBuild.  As part of the performance investigation/work we should go and talk to the folks that are bundling NuGet and ask them if they are OK with making this change to improve performance.

### Versions & Configurations
dotnet sdk 6.0.100-preview.2.21118.7

Binlog from when building the coreclr packages in dotnet/runtime which shows the `MissingMethodException`: 
[msbuild.zip](https://github.com/dotnet/msbuild/files/6031275/msbuild.zip)

--- a/src/MSBuild.UnitTests/Microsoft.Build.CommandLine.UnitTests.csproj
+++ b/src/MSBuild.UnitTests/Microsoft.Build.CommandLine.UnitTests.csproj
@@ -74,4 +74,10 @@
     <Service Include="{82a7f48d-3b50-4b1e-b82e-3ada8210c358}" />
   </ItemGroup>
 
+  <Target Name="CreateTaskDir" AfterTargets="Build" Condition="'$(TargetFrameworkIdentifier)' != ''">
+    <ItemGroup>
+      <OutputAssemblyList Include="$(TargetDir)Microsoft.Build.CommandLine.UnitTests.dll" />
+    </ItemGroup>
+    <Copy SourceFiles="@(OutputAssemblyList)" DestinationFolder="$(TargetDir)Task" />
+  </Target>
 </Project>

--- a/src/MSBuild.UnitTests/XMake_Tests.cs
+++ b/src/MSBuild.UnitTests/XMake_Tests.cs
@@ -2237,7 +2237,8 @@ namespace Microsoft.Build.UnitTests
 
 #if FEATURE_ASSEMBLYLOADCONTEXT
         /// <summary>
-        /// Ensure that tasks get loaded into their own <see cref="System.Runtime.Loader.AssemblyLoadContext"/>.
+        /// Ensure that tasks get loaded into their own <see cref="System.Runtime.Loader.AssemblyLoadContext"/>
+        /// if they are in a directory other than the MSBuild directory.
         /// </summary>
         /// <remarks>
         /// When loading a task from a test assembly in a test within that assembly, the assembly is already loaded
@@ -2247,7 +2248,10 @@ namespace Microsoft.Build.UnitTests
         [Fact]
         public void TasksGetAssemblyLoadContexts()
         {
-            string customTaskPath = Assembly.GetExecutingAssembly().Location;
+            string customTaskPath = Path.Combine(
+                Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location),
+                "Task",
+                Path.GetFileName(Assembly.GetExecutingAssembly().Location));
 
             string projectContents = $@"<Project ToolsVersion=`msbuilddefaulttoolsversion` xmlns=`msbuildnamespace`>
   <UsingTask TaskName=`ValidateAssemblyLoadContext` AssemblyFile=`{customTaskPath}` />
@@ -2259,7 +2263,6 @@ namespace Microsoft.Build.UnitTests
 
             ExecuteMSBuildExeExpectSuccess(projectContents);
         }
-
 #endif
 
         private string CopyMSBuild()

--- a/src/Shared/CoreCLRAssemblyLoader.cs
+++ b/src/Shared/CoreCLRAssemblyLoader.cs
@@ -23,7 +23,7 @@ namespace Microsoft.Build.Shared
 
         private bool _resolvingHandlerHookedUp = false;
 
-        private static string _msbuildDirPath;
+        private static readonly string _msbuildDirPath;
         private static readonly Version _currentAssemblyVersion = new Version(Microsoft.Build.Shared.MSBuildConstants.CurrentAssemblyVersion);
 
         static CoreClrAssemblyLoader()

--- a/src/Shared/CoreCLRAssemblyLoader.cs
+++ b/src/Shared/CoreCLRAssemblyLoader.cs
@@ -23,7 +23,13 @@ namespace Microsoft.Build.Shared
 
         private bool _resolvingHandlerHookedUp = false;
 
+        private static string _msbuildDirPath;
         private static readonly Version _currentAssemblyVersion = new Version(Microsoft.Build.Shared.MSBuildConstants.CurrentAssemblyVersion);
+
+        internal CoreClrAssemblyLoader()
+        {
+            _msbuildDirPath = Path.GetDirectoryName(BuildEnvironmentHelper.Instance.CurrentMSBuildExePath);
+        }
 
         public void AddDependencyLocation(string fullPath)
         {
@@ -52,7 +58,11 @@ namespace Microsoft.Build.Shared
             // folders in a NuGet package).
             fullPath = FileUtilities.NormalizePath(fullPath);
 
-            if (Traits.Instance.EscapeHatches.UseSingleLoadContext)
+            // If the requested load comes from the same directory as MSBuild, assume that
+            // the load is part of the platform, and load it using the Default ALC.
+            string assemblyDir = Path.GetDirectoryName(fullPath);
+
+            if (Traits.Instance.EscapeHatches.UseSingleLoadContext || string.Equals(assemblyDir, _msbuildDirPath))
             {
                 return LoadUsingLegacyDefaultContext(fullPath);
             }

--- a/src/Shared/CoreCLRAssemblyLoader.cs
+++ b/src/Shared/CoreCLRAssemblyLoader.cs
@@ -26,9 +26,9 @@ namespace Microsoft.Build.Shared
         private static string _msbuildDirPath;
         private static readonly Version _currentAssemblyVersion = new Version(Microsoft.Build.Shared.MSBuildConstants.CurrentAssemblyVersion);
 
-        internal CoreClrAssemblyLoader()
+        static CoreClrAssemblyLoader()
         {
-            _msbuildDirPath = FileUtilities.NormalizePath(BuildEnvironmentHelper.Instance.CurrentMSBuildExePath);
+            _msbuildDirPath = FileUtilities.NormalizePath(typeof(CoreClrAssemblyLoader).Assembly.Location);
             _msbuildDirPath = Path.GetDirectoryName(_msbuildDirPath);
         }
 
@@ -63,7 +63,8 @@ namespace Microsoft.Build.Shared
             // the load is part of the platform, and load it using the Default ALC.
             string assemblyDir = Path.GetDirectoryName(fullPath);
 
-            if (Traits.Instance.EscapeHatches.UseSingleLoadContext || string.Equals(assemblyDir, _msbuildDirPath))
+            if (Traits.Instance.EscapeHatches.UseSingleLoadContext ||
+                FileUtilities.ComparePathsNoThrow(assemblyDir, _msbuildDirPath, string.Empty))
             {
                 return LoadUsingLegacyDefaultContext(fullPath);
             }

--- a/src/Shared/CoreCLRAssemblyLoader.cs
+++ b/src/Shared/CoreCLRAssemblyLoader.cs
@@ -28,7 +28,8 @@ namespace Microsoft.Build.Shared
 
         internal CoreClrAssemblyLoader()
         {
-            _msbuildDirPath = Path.GetDirectoryName(BuildEnvironmentHelper.Instance.CurrentMSBuildExePath);
+            _msbuildDirPath = FileUtilities.NormalizePath(BuildEnvironmentHelper.Instance.CurrentMSBuildExePath);
+            _msbuildDirPath = Path.GetDirectoryName(_msbuildDirPath);
         }
 
         public void AddDependencyLocation(string fullPath)

--- a/src/Shared/MSBuildLoadContext.cs
+++ b/src/Shared/MSBuildLoadContext.cs
@@ -25,20 +25,8 @@ namespace Microsoft.Build.Shared
                 "MSBuild",
                 "Microsoft.Build",
                 "Microsoft.Build.Framework",
-                "Microsoft.Build.NuGetSdkResolver",
                 "Microsoft.Build.Tasks.Core",
                 "Microsoft.Build.Utilities.Core",
-                "NuGet.Build.Tasks",
-                "NuGet.Common",
-                "NuGet.Configuration",
-                "NuGet.Credentials",
-                "NuGet.DependencyResolver.Core",
-                "NuGet.Frameworks",
-                "NuGet.LibraryModel",
-                "NuGet.Packaging",
-                "NuGet.Protocol",
-                "NuGet.ProjectModel",
-                "NuGet.Versioning",
             }.ToImmutableHashSet();
 
         internal static readonly string[] Extensions = new[] { "ni.dll", "ni.exe", "dll", "exe" };

